core/arch/riscv/mm/core_mmu.c

```diff
@@ -0,0 +1,639 @@
+#include <kernel/spinlock.h>
+#include <kernel/tee_l2cc_mutex.h>
+#include <kernel/tee_misc.h>
+#include <kernel/tlb_helpers.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <riscv.h>
+#include <stdlib.h>
+#include <string.h>
+#include <trace.h>
+#include <util.h>
+
+#ifndef RV64
+#error implement
+#endif
+
+#ifndef DEBUG_XLAT_TABLE
+#define DEBUG_XLAT_TABLE 0
+#endif
+
+#if DEBUG_XLAT_TABLE
+#define debug_print(...) DMSG_RAW(__VA_ARGS__)
+#else
+#define debug_print(...) ((void)0)
+#endif
+
+bitstr_t g_asid __nex_bss;
+static unsigned int g_asid_spinlock __nex_bss = SPINLOCK_UNLOCK;
+
+typedef struct {
+	unsigned long entry;
+} pte_t;
+
+typedef struct {
+	pte_t entries[RISCV_PTES_PER_PT];
+} pgt_t;
+
+#define RISCV_MMU_PGT_SIZE	(RISCV_PTES_PER_PT * sizeof(pte_t))
+
+static pgt_t base_xlation_table
+	__aligned(RISCV_PGSIZE)
+	__section(".nozi.mmu.base_table");
+
+static pgt_t xlat_tables[RISCV_MMU_MAX_PGTS]
+	__aligned(RISCV_PGSIZE) __section(".nozi.mmu.l2");
+
+static pgt_t ta_tables[CFG_NUM_THREADS]
+	__aligned(RISCV_PGSIZE) __section(".nozi.mmu.l2");
+
+static int user_va_idx __nex_data = -1;
+
+struct mmu_partition {
+	pgt_t *base_table;
+	pgt_t *xlat_tables;
+	pgt_t *ta_tables;
+	unsigned int pgts_used;
+	unsigned int asid;
+};
+
+static struct mmu_partition default_partition __nex_data  = {
+	.base_table = &base_xlation_table,
+	.xlat_tables = xlat_tables,
+	.ta_tables = ta_tables,
+	.pgts_used = 0,
+	.asid = 0
+};
+
+static inline pte_t *core_mmu_table_get_entry(pgt_t *pgt, unsigned int idx)
+{
+	return &pgt->entries[idx & RISCV_MMU_VPN_MASK];
+}
+
+static void core_mmu_entry_set(pte_t *pte, uint64_t val)
+{
+	pte->entry = val;
+}
+
+static uint64_t core_mmu_entry_get(pte_t *pte)
+{
+	return pte->entry;
+}
+
+static inline bool core_mmu_entry_is_valid(pte_t *pte)
+{
+	return (pte->entry & PTE_V);
+}
+
+static inline bool core_mmu_entry_is_invalid(pte_t *pte)
+{
+	return !core_mmu_entry_is_valid(pte);
+}
+
+static inline bool core_mmu_entry_is_leaf(pte_t *pte)
+{
+	/* A leaf has one or more RWX bits set */
+	return (pte->entry & (PTE_R | PTE_W | PTE_X)) != 0;
+}
+
+static inline bool core_mmu_entry_is_branch(pte_t *pte)
+{
+	return !(core_mmu_entry_is_leaf(pte));
+}
+
+static inline unsigned long core_mmu_pte_create(unsigned long ppn, int type)
+{
+	return ((ppn << PTE_PPN_SHIFT) | PTE_V | type);
+}
+
+static inline unsigned long core_mmu_ptd_create(unsigned long ppn)
+{
+	return core_mmu_pte_create(ppn, PTE_V);
+}
+
+static unsigned long core_mmu_pte_ppn(pte_t *pte)
+{
+	return pte->entry >> PTE_PPN_SHIFT;
+}
+
+static unsigned long core_mmu_ppn(unsigned long addr)
+{
+	return (addr >> RISCV_PGSHIFT);
+}
+
+static unsigned long core_mmu_pte_pa(pte_t *pte)
+{
+	return (core_mmu_pte_ppn(pte) << RISCV_PGSHIFT);
+}
+
+static unsigned long core_mmu_pgt_to_satp(unsigned long asid, pgt_t *pgt)
+{
+	unsigned long satp = 0;
+	unsigned long pgt_ppn = ((unsigned long)pgt) >> RISCV_PGSHIFT;
```

Comment:
You're depending on `pgt` having the same physical as virtual address.

Reply:
Added `virt_to_phys(pgt)`